### DISCUSSION TITLE
How to put conditions in ferret?

### DISCUSSION THREAD

**Discussion by vartika102** [MontFerret/ferret#666] [Q&A] [answered]
I want to run a loop until a particular condition is satisfied, for which I might need to use a checking condition. How to do it? Also, how can I break a loop in between when a particular condition is not satisfied?

**Comment by ziflex** [comment]
> I want to run a loop until a particular condition is satisfied, for which I might need to use a checking condition. How to do it? Also, how can I break a loop in between when a particular condition is not satisfied?

Hi, Ferret's query language is based on Arango's Query Language (AQL) which is primarily declarative.
The only thing that allows you to do some kind of 'flow control' is ternary operator ``?`` in assignments.

In your case, though, I would try to use ``FILTER`` clause in your ``FOR IN`` loop. 
If it does not work for you, try to write your custom iterator that implements the logic. [Here is an example](https://github.com/MontFerret/ferret/blob/master/pkg/stdlib/html/pagination.go).

In the future releases, we are planning to [add ``WHILE`` loop](https://github.com/MontFerret/ferret/issues/262) in order to avoid scenarios when you have to implement custom solutions. But since it's not a part of the AQL spec, it was not added in the early releases.

**Comment by ziflex** [comment]
> I meet the same question.
> In my case I just use return None,fmt.errorf(xxx) to raise an error in my embeded version to stop all the ferret script runtime.

In embedded scenario, you can pass in a context with cancellation and terminate an execution at any time.

**Comment by ziflex** [comment, marked as answer]
> > > I meet the same question.
> > > In my case I just use return None,fmt.errorf(xxx) to raise an error in my embeded version to stop all the ferret script runtime.
> > 
> > 
> > In embedded scenario, you can pass in a context with cancellation and terminate an execution at any time.
> 
> 1.I have an udf in my case:
> fetch_navigate_url := func(ctx context.Context, args ...core.Value) (core.Value, error) {
> do you mean by cancel the ctx here?

No, I mean when you start an execution of your script, pass the context into ``Program.Run`` method:

```

	c := compiler.New()
	p := c.MustCompile(`
		FOR i IN 1..1000
		   RETURN i * 2
	`)

	ctx, cancel := context.WithCancel(context.Background())

	go func() {
			<-someChan
			cancel()
	}()

	p.Run(ctx)
```

> 2.Is there any sample way to pass and return complex Object(or pointor) between AQL script and compiler.FqlCompiler.RegisterFunction?
> 
> ```
> 
> type TableObject struct {
>        //columns,rows,headers,format,storage...
> }
> 
> func1 := func(ctx context.Context, args ...core.Value)(core.Value, error) ...
> func2 := func(ctx context.Context, args ...core.Value)(core.Value, error)...
> 
> //AQL script:
>       LET table_obj = func1(document.html)   //which returns TableObject 
>       LET ret_save = func2(table_obj)  // pass in TableObject
> ```

Yes, you can! Your type must implement [``core.Value``](https://github.com/MontFerret/ferret/blob/master/pkg/runtime/core/value.go#L11) interface to be used within runtime.
If you want to read properties of your type you must also implement [``core.Getter``](https://github.com/MontFerret/ferret/blob/master/pkg/runtime/core/value.go#L35) interface.

To iterate over your types you need to implement [``core.Iterable``](https://github.com/MontFerret/ferret/blob/master/pkg/runtime/core/value.go#L22) interface.

I'm gonna write a dedicated blog post about embedding Ferret and hopefully update the docs.
